Added pyproject.toml, dockerfile and docker compose

AnkushMalaker opened the pull request:
Made this mainly for my own use, but maybe slight changes can merge upstream.
Please close if not required. 

I used uv because of personal preference, and added the appropriate uv.lock.

AnkushMalaker commented:
I have been using this for a while and its actually running really smoothly!

danieljkemp commented:
Changing the build system should probably be in it's own PR

AnkushMalaker commented:
@danieljkemp Fair, can split into two PRs if there's interest in merging.

synesthesiam commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+    volumes:
+      - ./whisper-data:/data
+    environment:
+      - TZ=Asia/Kolkata
```
This should be removed or commented out.

synesthesiam commented on .dockerignore:
```diff
@@ -0,0 +1 @@
+.venv
```
Let's switch this to excluded format instead, so we ignore everything `*` and add just the files we want to include (with `!`).

synesthesiam commented on Dockerfile.gpu:
```diff
@@ -0,0 +1,19 @@
+FROM nvidia/cuda:12.6.3-cudnn-runtime-ubuntu22.04
+
+ENV UV_COMPILE_BYTECODE=1
```
Would it be better to just use the `--compile-bytecode` option in `uv`?

synesthesiam commented on Dockerfile.gpu:
```diff
@@ -0,0 +1,19 @@
+
+RUN --mount=type=cache,target=/root/.cache/uv \
+    --mount=type=bind,source=uv.lock,target=uv.lock \
+    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
```
Why bind mount this and `uv.lock` instead of copying them?

AnkushMalaker replied:
This again is what was mentioned for using docker + uv optimally. 
https://docs.astral.sh/uv/guides/integration/docker/#optimizations
This syntax was specifically copied from the following section: 
 https://docs.astral.sh/uv/guides/integration/docker/#intermediate-layers
 
 In hindsight, this is not required, but also like the other comment, is less verbose than others. 
 It is slightly more niche than the simple "copy" so we could keep the copy for simplicity instead of this. Your call!

synesthesiam commented on README.md:
````diff
@@ -26,13 +26,35 @@ script/run --model tiny-int8 --language en --uri 'tcp://0.0.0.0:10300' --data-di
     --model tiny-int8 --language en
 ```
 
+### GPU
````
We should document that this only works for an NVIDIA GPU

synesthesiam commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+      dockerfile: Dockerfile.gpu
+    command: [
+      "uv", "run", "wyoming-faster-whisper",
+      "--model", "large-v3",
```
The default model should be base

synesthesiam commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+    environment:
+      - TZ=Asia/Kolkata
+      - local_files_only=False
+      - HF_HUB_CACHE=/tmp
```
Should we map a volume for the hub cache so models aren't downloaded every time?

AnkushMalaker replied:
Based on some testing and reading the main.py file, --data-dir argument controls where everything is downloaded. Manually specifying HF_HUB_CACHE doesn't put the model in HF_HUB_CACHE atleast as long as --data-dir is provided. 
Since its already volume mounted in the docker compose file and also in the docker run, I think we don't need to mount hub cache.

synesthesiam commented on Dockerfile.gpu:
```diff
@@ -0,0 +1,19 @@
+
+ENV UV_COMPILE_BYTECODE=1
+
+COPY --from=ghcr.io/astral-sh/uv:latest /uv /uvx /bin/
```
Any specific reason to copy from astral-sh instead of installing `uv`?

AnkushMalaker commented:
@synesthesiam Hey, updated the PR addressing most comments. I've commented clarifications under some of them, feel free to ask for any further changes required. 
Instead of documenting that it only works for GPU I added another dockerfile for CPU. 
We could just as easily use the GPU dockerfile/docker-compose and simply comment `--device=cuda` and `deploy` sections in the compose, but the other solution is better (CPU docker image doesn't need cuda libs)

AnkushMalaker commented:
I built a docker image with the latest commit.
https://hub.docker.com/r/ankushm8/wyoming-faster-whisper

preimmortal commented on Dockerfile.gpu:
```diff
@@ -0,0 +1,17 @@
+RUN --mount=type=cache,target=/root/.cache/uv \
+    uv sync --frozen --compile-bytecode
+
+EXPOSE 10300
```
Should we add an entrypoint into the Dockerfile so that you don't need to specify the command during `docker run` / `docker-compose.yml`?

preimmortal commented on Dockerfile.gpu:
```diff
@@ -0,0 +1,17 @@
+RUN --mount=type=cache,target=/root/.cache/uv \
+    --mount=type=bind,source=uv.lock,target=uv.lock \
+    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
+    uv sync --frozen --no-install-project
```
I agree with the other comment about not changing the build system. 

If we want to do this we should be doing a multi-stage docker build to build the artifact (uv build), then copy the artifact into the final container.

It would help to avoid installing the uv binary inside of the application container.

preimmortal commented on README.md:
````diff
@@ -26,11 +26,47 @@ script/run --model tiny-int8 --language en --uri 'tcp://0.0.0.0:10300' --data-di
+#### Run Docker Image (GPU)
+
+``` sh
+docker run -it -p 10300:10300 -v /path/to/local/data:/data wyoming-whisper \
+    --model tiny-int8 --language en --device cuda --data-dir /path/to/local/data
````
Should add a note that nvidia-container-toolkit and cuda 12.6 drivers must be installed on the system
